[Step5] 자동차 경주(리팩터링)

AnKyungMoo opened the pull request:
리뷰 부탁드립니다~!

AnKyungMoo committed:
feat: 코드리뷰 반영

- 계산기 expressionContents 첫번째 값을 확장함수를 이용해서 확인하도록 변경
- 자동차경주 createCars를 map을 이용해서 자동차 리스트를 만들도록 변경
- printCurrentCarMovingDistance 문자열 템플릿 적용
- printMovingDistanceByTurn parameter명 변경
- turnOfPosition 을 객체 외부에서 접근할 때는 imutable 객체로 접근하도록 변경
- printWinners 요구사항에 맞게 출력형식 변경
- 테스트코드에서 Car 객체를 name을 기준으로 같은 객체인지 판별할 수 있게 변경

AnKyungMoo committed:
feat: Car의 hashCode를 name이 기준이 되게 변경

AnKyungMoo committed:
refactor: 로직 분리 및 패키지 변경

- 용도에 맞춰서 패키지 분리
- CarRacing에는 비즈니스 로직만 남기고 게임 실행 관련은 CarRacingExecute로 이동

AnKyungMoo committed:
feat: 코드 리뷰 반영하면서 코드 수정

- car 객체의 go 함수가 다른 객체의 의존성이 약해지게 변경
- 차 이름 글자 갯수제한 추가
- 테스트코드에서 차 위치를 직접 이동시키지 않고 함수를 통해 이동하도록 변경

AnKyungMoo committed:
test: 차 글자 갯수 제한 체크하는 테스트 추가

karian7 commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -11,7 +11,7 @@ class Calculator {
         val expressionContents: List<String> = expression.split(" ")
 
-        var result = expressionContents[0].toInt()
+        var result = expressionContents.first().toInt()
```
👍

karian7 commented on src/main/kotlin/racing/domain/CarRacing.kt:
```diff
@@ -0,0 +1,27 @@
+import racing.dto.Car
+import racing.dto.Winners
+
+class CarRacing {
```
좋은 객체지향설계에서 객체 특히 도메인 객체는 "상태", "행동", "식별자" 를 가진다고 합니다.
https://it-mesung.tistory.com/100

`CarRacing` 은 도메인에서 필요한 로직을 담은 함수를 모아놓은 형태로 보입니다. 모든 메소드들이 static 으로 되어 있어도 크게 이상하지 않고 다시말해 객체지향적이지 않아 보여요.

객체에 상태를 주고 상태에 따라 행동하도록 바꿔봅시다.

karian7 commented on src/main/kotlin/racing/domain/CarRacing.kt:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    fun pickWinners(cars: List<Car>): Winners {
+        val maxDistance: Int = cars.maxOf { car -> car.turnOfPosition.last() }
+        val winnerNames: List<String> = cars.filter { car ->
+            car.turnOfPosition.last() >= maxDistance
+        }.map { car ->
+            car.name
+        }
```
우승자를 계산하는 책임을 `Winners`에게 옮기면 어떨까요? 어떤 장점이 있을까요.

AnKyungMoo replied:
테스트 코드를 작성할 때, 자동차 경주에 관련된 기능들을 가진 CarRacing 객체를 가져오지 않고
Winner 객체를 통해 우승자를 계산하는 부분만 확인할 수 있게 되네요!

karian7 commented on src/main/kotlin/racing/domain/CarRacing.kt:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    fun createCars(carNames: List<String>): List<Car> {
+        return carNames.map { carName -> Car(carName) }
```
```suggestion
        return carNames.map { Car(name = it) }
```
이건 어떠신가요?

AnKyungMoo replied:
훨씬 간결한데 가독성은 더 좋아보이네요!
it을 무조건적으로 지향하기보다는 네임드파라미터를 이용하면 더 가독성 있는 코드가 될 수 있군요..! 의견 감사합니다!

karian7 commented on src/main/kotlin/racing/dto/Car.kt:
```diff
@@ -0,0 +1,38 @@
+
+    private val _turnOfPosition = arrayListOf(START_POSITION)
+    val turnOfPosition: List<Int>
+        get() = _turnOfPosition
```
뒷받침 하는 프러퍼티 활용 👍

karian7 commented on src/main/kotlin/racing/dto/Car.kt:
```diff
@@ -0,0 +1,38 @@
+        return if ((0..9).random() >= REQUIRE_GO_VALUE) GO else STAY
+    }
+
+    fun go(goAmount: () -> Int = { goIfAboveRequireGoValue() }) {
```
Nit: 이런 표현도 가능합니다.
```suggestion
    fun go(goAmount: () -> Int = ::goIfAboveRequireGoValue) {
```

karian7 commented on src/main/kotlin/racing/view/CarRacingInputView.kt:
```diff
@@ -1,12 +1,21 @@
     }
+
+    fun checkCarNameLength(carName: String) {
+        check(carName.length <= 5) { "차 이름은 5글자를 초과 할 수 없습니다." }
```
```suggestion
        check(carName.length <= 5) { "차 이름은 5글자 이하여야 합니다." }
```
사소한 내용이지만 check 의 조건과 메시지가 일치하는것이 더 읽기 쉽다고 생각합니다.

karian7 commented on src/main/kotlin/racing/view/CarRacingInputView.kt:
```diff
@@ -1,12 +1,21 @@
         println("경주할 자동차 이름을 입력하세요(이름은 쉼표(,)를 기준으로 구분).")
         val carNames: List<String> = readLine()!!.split(",")
+
+        carNames.forEach(::checkCarNameLength)
```
자동차 이름에 대한 조건은 도메인 요구사항 입니다. 도메인 요구사항은 도메인 클래스에 포함되도록 개선해 봅시다. 이렇게 바꾸면 어떤 장점이 있을까요?

karian7 commented on src/main/kotlin/racing/view/CarRacingResultView.kt:
```diff
@@ -0,0 +1,32 @@
+    fun printWinners(winners: Winners) {
+        winners.winnerNames.dropLast(1).forEach { winnerName ->
+            print("$winnerName, ")
+        }
```
아래 내용을 확인해 보고 개선해 봅시다.

https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/join-to-string.html

karian7 commented on src/test/kotlin/racing/CarRacingTest.kt:
```diff
@@ -14,75 +19,77 @@ class CarRacingTest {
-        val carC = Car("c").apply { repeat(moveCount) { turnOfPosition.add(1) } }
-        val carD = Car("d").apply { repeat(moveCount) { turnOfPosition.add(2) } }
-        val carE = Car("e").apply { repeat(moveCount) { turnOfPosition.add(2) } }
+        val carA = Car("a").also { it.go { 2 } }
```
```suggestion
        val carA = Car("a", position = 2)
```
https://github.com/next-step/kotlin-racingcar/pull/526#discussion_r872564317 에 대한 추가 피드백입니다. 위 코드로 개선해 볼까요?

AnKyungMoo replied:
생성자에서 position = 2를 두는 것은 오로지 테스트를 위함인가요?

아니면 default 위치를 유연하게 하기 위함인가요?

karian7 replied:
테스트를 위해서 입니다.

karian7 commented on src/test/kotlin/racing/CarRacingTest.kt:
```diff
@@ -14,75 +19,77 @@ class CarRacingTest {
+    fun `차 이름이 5글자를 초과하여 IllegalStateException을 발생시킨다`() {
+        assertThrows<IllegalStateException> {
+            CarRacingInputView.checkCarNameLength("나비보벳따우")
+        }
```
만약 이 로직에 다른곳으로 옮겨진다면 테스트도 옮겨져야 할 것 같아요.

karian7 commented on src/main/kotlin/racing/view/CarRacingResultView.kt:
```diff
@@ -0,0 +1,32 @@
+        }
+    }
+
+    private fun printMovingDistanceByTurn(cars: List<Car>, turn: Int) {
```
```suggestion
    private inline fun printMovingDistanceByTurn(cars: List<Car>, positionOf: (Car) -> Int) {
```
변수의 전달범위를 조절하면 가독성이 올라가기도 합니다. 다만 함수형 프로그래밍에 익숙하지 않다면 오히려 방해가 되니 적절히 사용해야 해요. 브랜은 어떠신가요.

관련 피드백: https://github.com/next-step/kotlin-racingcar/pull/526#discussion_r872550955

AnKyungMoo replied:
해당 부분이 익숙한 모습이 아니어서 그런 것인지.. 
어떤 점이 가독성을 올리는 것인지 이해가 어렵습니다ㅜㅜ..

```
private fun printMovingDistanceByTurn(cars: List<Car>, turn: Int) {
    cars.forEach { car ->
        printCurrentCarMovingDistance(carName = car.name, carPosition = car.turnOfPosition[turn])
    }
    println()
}
```

car의 속해있는 정보들을 이용해서 printCurrentCarMovingDistance() 함수를 호출하고 있습니다.
position도 name과 같은 방식으로 파라미터로 전달하는 편이 같은 객체내의 정보를 전달한다고 직관적으로 표현하는 것이 아닐까 생각이 드네요..!

의견 주시면 감사하겠습니다!!

AnKyungMoo replied:
계속 고민하다보니 결국에는 car 객체를 넘겨서 차이가 없군요..!!
제가 질문하면서 글을 적다가 정리가 되었습니다

```
private inline fun printMovingDistanceByTurn(cars: List<Car>, positionOf: (Car) -> Int) {
    cars.forEach { car ->
        printCurrentCarMovingDistance(carName = car.name, carPosition = positionOf(car))
    }
    println()
}
```

karian7 replied:
네 ㅎㅎ https://github.com/next-step/kotlin-racingcar/pull/580#discussion_r873574071 에서도 말씀드린것처럼 조직내 숙련도에 따라 더 안좋을 수도 있어요. for vs forEach 같이요

karian7 reviewed (changes requested):
마지막 단계 잘 수행해 주셨습니다. 👏 
밤 늦게까지 만들어 주셨는데 오늘은 피드백이 좀 늦었네요 😓 

요구사항에 명시되지 않았지만 사용자 입력값 (carNames, moveCount) 에 대한 추가 검증과 단위테스트를 만들어 보면 좋겠습니다. 더불어 이 제약조건들은 도메인 내에 위치하도록 고민해 주세요.

그럼 검토해 보시고 다시 요청 부탁드려요~!

AnKyungMoo committed:
feat: 코드리뷰 반영

- CarRacing에 상태 부여
- 우승자 선택하는 함수를 Winner로 이동
- 이름 길이 체크하는 함수의 위치 변경
- 결과 출력방식 변경
- 테스트를 위해 차의 초기 위치 설정하는 코드 추가

karian7 commented on src/main/kotlin/racing/CarRacingExecute.kt:
```diff
@@ -1,5 +1,26 @@
+
+    fun startRacing() {
+        val (carNames, moveCount) = CarRacingInputView.input()
+        carNames.forEach { carRacing.checkCarNameLength(carName = it) }
```
```suggestion
        val carRacing = CarRacing(carNames)
```

karian7 commented on src/main/kotlin/racing/domain/CarRacing.kt:
```diff
@@ -0,0 +1,24 @@
+        _cars.addAll(carNames.map { Car(name = it) })
+    }
+
+    fun checkCarNameLength(carName: String) {
+        check(carName.length <= 5) { "차 이름은 5글자 이하여야 합니다." }
+    }
```
자동차 이름의 대한 조건은 `Car` 객체에서 검증하는게 좋을것 같습니다.
이렇게 하면 생성된 모든 `Car` 객체는 이름규약을 지키고 있는것으로 간주해도 되기 때문입니다.
두번째로는 다른 객체에 대한 의존없이 `Car` 객체는 `CarTest` 에서 테스트 할 수 있습니다.

karian7 commented on src/test/kotlin/racing/CarRacingTest.kt:
```diff
@@ -2,87 +2,83 @@ package racing
+        val winners = Winners(cars = cars).pickWinners()
         assertThat(winners.size).isEqualTo(1)
-        assertThat(winners[0]).isEqualTo("c")
+        assertThat(winners[0].name).isEqualTo("c")
```
`Winners`에 대한 테스트는 `WinnersTest.kt` 에 작성하면 좋을것 같습니다.

karian7 reviewed (approved):
수많은 피드백들 반영하시느라 고생 많으셨습니다. 💯 
적극적인 질문을 해 주셔서 저도 잘 설명하기 위해 공부할 수 있었습니다.

이번 미션에서는 단위테스트를 `CarRacingTest` 에 모두 만드셨는데요, 다음 미션에서는 각각의 요구사항들을 책임지는 도메인 객체를 식별하고 각 도메인 객체에 대한 테스트를 만들어 보세요. 다양한 단위테스트를 만드는 기법을 알게 될 것입니다.

다음 미션도 화이팅이고 부디 마지막 미션까지 완주하시길 빕니다 🙏 
자동차 미션 리뷰는 마무리 하겠습니다! 🚗 🏁🏁🏁
